src/vcpkg/base/cmd-parser.cpp

```diff
@@ -82,4 +324,692 @@ namespace vcpkg
+        while (first != last)
+        {
+            if (first->empty() || first->front() != '@')
+            {
+                ++first;
+                continue;
+            }
+
+            const auto file_name = static_cast<StringView>(*first);
+            auto maybe_response_file_lines = response_file_source.read_lines(file_name.substr(1));
+            if (auto response_file_lines = maybe_response_file_lines.get())
+            {
+                if (response_file_lines->empty())
+                {
+                    first = inputs.erase(first);
+                    last = inputs.end();
+                    continue;
+                }
+
+                // replace the response file name with the first line, and insert the rest
+                auto lines_begin = std::make_move_iterator(response_file_lines->begin());
+                const auto lines_end = std::make_move_iterator(response_file_lines->end());
+                *first = *lines_begin;
+                ++first;
+                ++lines_begin;
+                if (lines_begin != lines_end)
+                {
+                    first = inputs.insert(first, lines_begin, lines_end);
+                    first += response_file_lines->size() - 1;
+                    last = inputs.end();
+                }
+            }
+            else
+            {
+                return std::move(maybe_response_file_lines).error();
+            }
+        }
+
+        return Unit{};
+    }
+
+    bool SwitchName::operator<(const SwitchName& rhs) const { return switch_name < rhs.switch_name; }
+
+    CmdParser::CmdParser()
+        : argument_strings(), argument_strings_lowercase(), argument_parsed(), errors(), options_table()
+    {
+    }
+
+    CmdParser::CmdParser(View<std::string> inputs)
+        : argument_strings(inputs.begin(), inputs.end())
+        , argument_strings_lowercase()
+        , argument_parsed(argument_strings.size(), false)
+        , errors()
+        , options_table()
+    {
+        insert_lowercase_strings(argument_strings_lowercase, argument_strings);
+    }
+
+    CmdParser::CmdParser(std::vector<std::string>&& inputs)
+        : argument_strings(std::move(inputs))
+        , argument_strings_lowercase()
+        , argument_parsed(argument_strings.size(), false)
+        , errors()
+        , options_table()
+    {
+        insert_lowercase_strings(argument_strings_lowercase, argument_strings);
+    }
+
+    CmdParser::CmdParser(const CmdParser&) = default;
+    CmdParser::CmdParser(CmdParser&&) = default;
+    CmdParser& CmdParser::operator=(const CmdParser&) = default;
+    CmdParser& CmdParser::operator=(CmdParser&&) = default;
+
+    bool CmdParser::parse_switch(StringView switch_name, StabilityTag stability, bool& value)
+    {
+        std::size_t found = 0;
+        for (std::size_t idx = 0; idx < argument_strings.size(); ++idx)
+        {
+            if (argument_parsed[idx] != false)
+            {
+                continue;
```

Comment:
Completely skipping 'used' arguments makes me nervous -- could you explain where this behavior is helpful?

Reply:
This was necessary when I allowed options to have values-that-look-like-options.

That is:

```
CmdParser uut{{"--option", "--value"}};
assert(uut.parse_option("option") == "--value");
assert(uut.parse_switch("value") == false);
```

Now that we decided that this is no longer allowed I think this check is removable, at the expense of potentially doing strange things on misuse. That is:

```
CmdParser uut{{"--switch"}};
assert(uut.parse_switch("switch") == true);
assert(uut.parse_switch("switch") == false);
```

succeeds today but would fail if I removed the check. This is tested on line 207 of cmd-parser.cpp (in the unit tests directory)

In contrast, for the "options" and "multi_options" the check is still necessary. Consider:

```
CmdParser uut{{"--option", "a", "b"}};
assert(uut.extract_first_command_like_arg_lowercase().value_or_exit(VCPKG_LINE_INFO) == "a");
assert(!uut.parse_option("option"));
```

Without knowing that "a" was already consumed we wouldn't detect this error.